staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go

```diff
@@ -90,7 +90,14 @@ func customMarshalValue(value reflect.Value) (reflect.Value, bool) {
 
 	marshaler, ok := value.Interface().(Marshaler)
 	if !ok {
-		return reflect.Value{}, false
+		if !isPointerKind(value.Kind()) && value.CanAddr() {
```

Comment:
Previously, a call to `marshal(metav1.Now())` would marshal correctly.  Wouldn't the `CanAddr` fail now?  Worth adding a test since I'm sure someone else will have the question.

Reply:
@ash2k I'm not sure we all came to a mutual understanding on this point. Do you guys agree that this is true:

>If only the pointer type implements Marshaler, passing a value type to customMarshalValue() causes CanAddr to return false.

And if so, do you agree that's a change from previous behavior, e.g.:

>Previously, a call to marshal(metav1.Now()) would marshal correctly. 

And if that's also true, is all that okay?

Reply:
@ironcladlou @deads2k What `marshal()` are we talking about? 

`MarshalQueryParameter()` is a method on `metav1.Time`. It is only ever called via reflection from [`customMarshalValue()`](https://github.com/atlassian/kubernetes/blob/6ec536513a4c85af08f43def4bea1fdddde9f215/staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go#L85) which is only ever called from [`convertStruct()`](https://github.com/atlassian/kubernetes/blob/6ec536513a4c85af08f43def4bea1fdddde9f215/staging/src/k8s.io/apimachinery/pkg/conversion/queryparams/convert.go#L162) which is only called for structs. So `metav1.Time` is always a field in a struct and it should always be possible to take the address of a field.

However I agree that if there was a method `marshal()` that tried to call `MarshalQueryParameter()` directly on a value (via reflection) and that value was `metav1.Time` then yes, it would be a change in behavior because it will not work (because of a pointer receiver). However I think this is a reasonable requirement to ask to pass a pointer to a value not just the value. That is how `encoding/json` package works: https://play.golang.org/p/jKkAXSJ9Ul

In the example above `MarshalJSON()` method is not picked up by `encoding/json` for the value/value case.

p.s. I was confused and now I understand that initial comment.

Reply:
I'm actually happy to remove the change to the method receiver. The original issue has been fixed in #45167 anyway and I've audited all other places where `MarshalQueryParameter()` is called and they are not using a pointer `Time`.
I still believe the fix for reflection check should be merged though. WDYT?

Reply:
^ done.

Reply:
Thanks for the additional explanation... I agree that requiring a pointer argument is a reasonable requirement, and your detail about field addressability makes sense. I *think* everything makes sense now and looks good, and wish somebody else would chime in, but LGTM. :grin: